crate/server/src/routes/google_cse/operations.rs

```diff
@@ -193,7 +220,261 @@ pub async fn unwrap(
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub rsa_oaep_label: Option<String>,
+    /// A passthrough JSON string providing additional context about the operation. The JSON provided should be sanitized before being displayed. Max size: 1 KB.
+    pub reason: String,
+    /// The base64-encoded wrapped private key. Max size: 8 KB.
+    pub wrapped_private_key: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct PrivateKeyDecryptResponse {
+    /// A base64-encoded data encryption key.
+    pub data_encryption_key: String,
+}
+
+/// Unwraps a wrapped private key and then decrypts the content encryption key that is encrypted to the public key.
+///
+/// See Google documentation:
+/// - Private Key Decrypt endpoint: https://developers.google.com/workspace/cse/reference/private-key-decrypt
+pub async fn private_key_decrypt(
+    req_http: HttpRequest,
+    request: PrivateKeyDecryptRequest,
+    cse_config: &Arc<Option<GoogleCseConfig>>,
+    kms: &Arc<KMSServer>,
+) -> KResult<PrivateKeyDecryptResponse> {
+    debug!("private_key_decrypt: entering");
+    let database_params = kms.get_sqlite_enc_secrets(&req_http)?;
+
+    debug!("private_key_decrypt: validate_tokens");
+    let user = validate_tokens(
+        &request.authentication,
+        &request.authorization,
+        cse_config,
+        "gmail",
+        None,
+    )
+    .await?;
+
+    debug!("private_key_decrypt: check algorithm");
+    kms_ensure!(
+        request.algorithm == "RSA/ECB/PKCS1Padding",
+        "Only RSA/ECB/PKCS1Padding is supported"
+    );
+
+    // Base 64 decode the encrypted DEK and create a wrapped KMIP object from the key bytes
+    debug!("private_key_decrypt: decode encrypted_dek");
+    let encrypted_dek = general_purpose::STANDARD.decode(&request.encrypted_data_encryption_key)?;
+
+    // Unwrap private key which has been previously wrapped using AES
+    let dek = cse_symmetric_unwrap(request.wrapped_private_key, user, kms, database_params).await?;
+
+    // Decrypt with the unwrapped RSA private key
+    debug!("private_key_decrypt: from_rsa");
+    let private_key = PKey::from_rsa(Rsa::<Private>::private_key_from_der(&dek)?)?;
+
+    // Perform RSA PKCS1 decryption.
+    let mut ctx = PkeyCtx::new(&private_key)?;
+    ctx.decrypt_init()?;
+    ctx.set_rsa_padding(Padding::PKCS1)?;
+    if let Some(label) = request.rsa_oaep_label {
+        ctx.set_rsa_oaep_label(label.as_bytes())?;
+        ctx.set_rsa_padding(Padding::PKCS1_OAEP)?;
+    }
+    let allocation_size = ctx.decrypt(&encrypted_dek, None)?;
+    debug!("privatekeydecrypt: allocation_size: {allocation_size}");
+    let mut plaintext = vec![0_u8; allocation_size];
+    let decrypt_size = ctx.decrypt(&encrypted_dek, Some(&mut *plaintext))?;
+
+    kms_ensure!(
+        decrypt_size > 0,
+        "privatekeydecrypt: decrypted plaintext cannot be empty"
+    );
```

Comment:
Why do we need this specific check ? We should check the lenght of the encrypted text instead no? If plaintext is empty - it means that encrypted text was empty first